src/dpp/voice/enabled/handle_frame.cpp

```diff
@@ -0,0 +1,450 @@
+				}
+				mls_state->encryptor->set_key_ratchet(mls_state->dave_session->get_key_ratchet(creator->me.id.str()));
+
+				/**
+				 * https://www.ietf.org/archive/id/draft-ietf-mls-protocol-14.html#name-epoch-authenticators
+				 * 9.7. Epoch Authenticators
+				 * The main MLS key schedule provides a per-epoch epoch_authenticator. If one member of the group is being impersonated by an active attacker,
+				 * the epoch_authenticator computed by their client will differ from those computed by the other group members.
+				 */
+				mls_state->privacy_code = generate_displayable_code(mls_state->dave_session->get_last_epoch_authenticator());
+				log(ll_debug, "E2EE Privacy Code: " + mls_state->privacy_code);
+
+				if (!creator->on_voice_ready.empty()) {
+					voice_ready_t rdy(nullptr, data);
+					rdy.voice_client = this;
+					rdy.voice_channel_id = this->channel_id;
+					creator->on_voice_ready.call(rdy);
+				}
+			}
+			break;
+			case voice_client_dave_mls_welcome: {
+				this->mls_state->transition_id = dave_header->get_welcome_transition_id();
+				log(ll_debug, "voice_client_dave_mls_welcome with transition id " + std::to_string(this->mls_state->transition_id));
+				auto r = mls_state->dave_session->process_welcome(dave_header->get_welcome_data(data.length()), dave_mls_user_list);
+				if (r.has_value()) {
+					for (const auto& user : dave_mls_user_list) {
+						log(ll_debug, "Setting decryptor key ratchet for user: " + user + ", protocol version: " + std::to_string(mls_state->dave_session->get_protocol_version()));
+						dpp::snowflake u{user};
+						mls_state->decryptors.emplace(u, std::make_unique<dpp::dave::decryptor>());
+						mls_state->decryptors.find(u)->second->transition_to_key_ratchet(mls_state->dave_session->get_key_ratchet(user));
+					}
+					mls_state->encryptor->set_key_ratchet(mls_state->dave_session->get_key_ratchet(creator->me.id.str()));
+				}
+				mls_state->privacy_code = generate_displayable_code(mls_state->dave_session->get_last_epoch_authenticator());
+				log(ll_debug, "E2EE Privacy Code: " + mls_state->privacy_code);
+
+				json obj = {
+					{ "op", voice_client_dave_transition_ready },
+					{
+					  "d",
+						{
+							{ "transition_id", this->mls_state->transition_id },
+						}
+					}
+				};
+				this->write(obj.dump(-1, ' ', false, json::error_handler_t::replace), OP_TEXT);
+			}
+			break;
+			default:
+				log(ll_debug, "Unexpected DAVE frame opcode");
+				log(dpp::ll_trace, "R: " + dpp::utility::debug_dump((uint8_t*)(data.data()), data.length()));
+			break;
+		}
+
+		return true;
+	}
+
+	try {
+		log(dpp::ll_trace, std::string("R: ") + data);
+		j = json::parse(data);
+	}
+	catch (const std::exception &e) {
+		log(dpp::ll_error, std::string("discord_voice_client::handle_frame ") + e.what() + ": " + data);
+		return true;
+	}
+
+	if (j.find("seq") != j.end() && j["seq"].is_number()) {
```

Comment:
So many unnecessary lookups! Every one of these `j["seq"]` searches through the json object... You can use the iterator returned by `find`, since you've established it is valid

Reply:
this has always been this way. do we have to change it now?